File: samples/dnn/paddleocr_vl_inference.py
# This file is part of OpenCV project.
# It is subject to the license terms in the LICENSE file found in the top-level directory
# of this distribution and at http://opencv.org/license.html.


'''
This is a sample script to run PaddleOCR-VL-1.5 vision-language inference in OpenCV
using ONNX models and OpenCV's new DNN engine (ENGINE_NEW). Given an image and a text
prompt, it generates a text response (e.g. recognized document text).

The model is split into three ONNX files:
    - Vision encoder : image patches -> image embeddings
    - Embedding      : prompt token ids -> text embeddings
    - Decoder        : [image embeddings | text embeddings] -> logits (with KV-cache)

Model: https://huggingface.co/PaddlePaddle/PaddleOCR-VL
ONNX:  https://huggingface.co/onnx-community/PaddleOCR-VL-1.5-ONNX

Run the script:
1. Download the plain (non-quantized) fp32 ONNX export into <model_dir>, keeping the
   upstream layout (config.json, tokenizer.json, processor_config.json at the root;
   onnx/vision_encoder.onnx, onnx/embedding.onnx, onnx/decoder.onnx under onnx/):

2. Run the script:

    python paddleocr_vl_inference.py --model_dir=<model_dir> \
                                     --input=<path-to-image>
'''

import json
import argparse
import numpy as np
import cv2 as cv

def parse_args():
    parser = argparse.ArgumentParser(description='Use this script to run PaddleOCR-VL-1.5 vision-language inference in OpenCV',
                                    formatter_class=argparse.ArgumentDefaultsHelpFormatter)
    parser.add_argument('--model_dir', type=str, required=True,
                        help='Path to the local onnx-community/PaddleOCR-VL-1.5-ONNX export '
                             '(config.json, processor_config.json, '
                             'onnx/{vision_encoder,embedding,decoder}.onnx).')
    parser.add_argument('--input', '-i', type=str, required=True, help='Path to the input image.')
    parser.add_argument('--prompt', type=str, default='OCR', help='Task prompt.')
    parser.add_argument('--max_new_tokens', type=int, default=512, help='Maximum number of new tokens to generate.')
    parser.add_argument('--seed', type=int, default=0, help='Random seed.')
    return parser.parse_args()

def load_json(path):
    with open(path, 'r', encoding='utf-8') as f:
        return json.load(f)

def config_value(config, name, default=None):
    value = config.get(name)
    if value is not None:
        return value
    return (config.get('text_config') or {}).get(name, default)

def smart_resize(height, width, factor, min_pixels, max_pixels):
    '''Qwen2VL-style resize: round to a multiple of factor while keeping the pixel budget
    within [min_pixels, max_pixels].'''
    if height < factor:
        width = round((width * factor) / height)
        height = factor
    if width < factor:
        height = round((height * factor) / width)
        width = factor
    if max(height, width) / min(height, width) > 200:
        raise ValueError(f'absolute aspect ratio is too large: {height}x{width}')

    h_bar = round(height / factor) * factor
    w_bar = round(width / factor) * factor
    if h_bar * w_bar > max_pixels:
        beta = np.sqrt((height * width) / max_pixels)
        h_bar = int(np.floor(height / beta / factor) * factor)
        w_bar = int(np.floor(width / beta / factor) * factor)
    elif h_bar * w_bar < min_pixels:
        beta = np.sqrt(min_pixels / (height * width))
        h_bar = int(np.ceil(height * beta / factor) * factor)
        w_bar = int(np.ceil(width * beta / factor) * factor)
    return h_bar, w_bar

def preprocess_image(image_path, preprocessor):
    '''Smart-resize then flatten into merge_size x merge_size patches, matching the
    Qwen2VL-style image processor this vision encoder was exported with.'''
    patch_size = int(preprocessor.get('patch_size', 14))
    merge_size = int(preprocessor.get('merge_size', 2))
    temporal_patch_size = int(preprocessor.get('temporal_patch_size', 1))
    min_pixels = int(preprocessor.get('min_pixels', 28 * 28 * 130))
    max_pixels = int(preprocessor.get('max_pixels', 28 * 28 * 1280))
    image_mean = np.asarray(preprocessor.get('image_mean', [0.5, 0.5, 0.5]), dtype=np.float32)
    image_std = np.asarray(preprocessor.get('image_std', [0.5, 0.5, 0.5]), dtype=np.float32)
    rescale_factor = float(preprocessor.get('rescale_factor', 1 / 255))

    image = cv.imread(image_path)
    if image is None:
        raise IOError('Could not read image: ' + image_path)
    height, width = image.shape[:2]
    resized_height, resized_width = smart_resize(height, width, patch_size * merge_size, min_pixels, max_pixels)
    image = cv.resize(image, (resized_width, resized_height), interpolation=cv.INTER_CUBIC)
    image = cv.cvtColor(image, cv.COLOR_BGR2RGB).astype(np.float32) * rescale_factor
    image = (image - image_mean.reshape(1, 1, 3)) / image_std.reshape(1, 1, 3)

    patches = image.transpose(2, 0, 1)[np.newaxis]
    if patches.shape[0] == 1:
        patches = np.tile(patches, (temporal_patch_size, 1, 1, 1))

    channel = patches.shape[1]
    grid_t = patches.shape[0] // temporal_patch_size
    grid_h = resized_height // patch_size
    grid_w = resized_width // patch_size
    patches = patches.reshape(grid_t, temporal_patch_size, channel, grid_h, patch_size, grid_w, patch_size)
    patches = patches.transpose(0, 3, 5, 2, 1, 4, 6)
    flatten_patches = patches.reshape(grid_t * grid_h * grid_w, channel, patch_size, patch_size).astype(np.float32, copy=False)
    pixel_values = flatten_patches[np.newaxis]
    image_grid_thw = np.asarray([[grid_t, grid_h, grid_w]], dtype=np.int64)
    return pixel_values, image_grid_thw

def build_prompt(prompt, image_token_repeats):
    return ('<|begin_of_sentence|>User: <|IMAGE_START|>'
            + '<|IMAGE_PLACEHOLDER|>' * image_token_repeats
            + '<|IMAGE_END|>' + prompt + '\nAssistant:\n')

def paddleocr_vl_inference(vision_net, embed_net, decoder_net, pixel_values, image_grid_thw, prompt,
                           max_new_tokens, tokenizer, image_token_id, eos_token_id, merge_size):

    print('Inferencing PaddleOCR-VL-1.5 model...')

    image_token_repeats = int(np.prod(image_grid_thw[0]) // merge_size // merge_size)
    input_ids = np.array([tokenizer.encode(build_prompt(prompt, image_token_repeats))], dtype=np.int64)
    prompt_len = input_ids.shape[1]

    vision_net.setInput(pixel_values, 'pixel_values')
    vision_net.setInput(image_grid_thw, 'image_grid_thw')
    image_embeds = vision_net.forward()

    embed_net.setInput(input_ids, 'input_ids')
    inputs_embeds = embed_net.forward()
    image_positions = input_ids == image_token_id
    inputs_embeds[image_positions] = image_embeds.reshape(-1, image_embeds.shape[-1])

    decoder_net.enableKVCache()
    attention_mask = np.ones((1, prompt_len), dtype=np.int64)

    decoder_net.setInput(inputs_embeds, 'inputs_embeds')
    decoder_net.setInput(attention_mask, 'attention_mask')
    logits = decoder_net.forward()
    new_id = int(np.argmax(logits[:, -1, :].reshape(-1)))
    generated = [new_id]

    for _ in range(max_new_tokens - 1):
        if new_id == eos_token_id:
            break
        embed_net.setInput(np.array([[new_id]], dtype=np.int64), 'input_ids')
        new_embed = embed_net.forward()
        attention_mask = np.concatenate([attention_mask, [[1]]], axis=-1)
        decoder_net.setInput(new_embed, 'inputs_embeds')
        decoder_net.setInput(attention_mask, 'attention_mask')
        logits = decoder_net.forward()
        new_id = int(np.argmax(logits[:, -1, :].reshape(-1)))
        generated.append(new_id)

    if generated and generated[-1] == eos_token_id:
        generated.pop()

    return generated

if __name__ == '__main__':

    args = parse_args()
    np.random.seed(args.seed)

    print('Preparing PaddleOCR-VL-1.5 model...')
    tokenizer = cv.dnn.Tokenizer.loadVLM(args.model_dir + '/', 'paddleocr-vl')

    config = load_json(args.model_dir + '/config.json')
    processor = load_json(args.model_dir + '/processor_config.json')['image_processor']
    image_token_id = int(config_value(config, 'image_token_id'))
    eos_token_id = int(config_value(config, 'eos_token_id', 2))
    merge_size = int(processor.get('merge_size', 2))

    vision_net  = cv.dnn.readNetFromONNX(args.model_dir + '/onnx/vision_encoder.onnx', cv.dnn.ENGINE_NEW)
    embed_net   = cv.dnn.readNetFromONNX(args.model_dir + '/onnx/embedding.onnx', cv.dnn.ENGINE_NEW)
    decoder_net = cv.dnn.readNetFromONNX(args.model_dir + '/onnx/decoder.onnx', cv.dnn.ENGINE_NEW)

    pixel_values, image_grid_thw = preprocess_image(args.input, processor)
    print(f'Prompt:\n{args.prompt}')

    generated = paddleocr_vl_inference(vision_net, embed_net, decoder_net, pixel_values, image_grid_thw,
                                       args.prompt, args.max_new_tokens, tokenizer, image_token_id,
                                       eos_token_id, merge_size)
    response = tokenizer.decode(generated)
    print(f'Response:\n{response}')

File: samples/dnn/paddleocr_vl_inference.cpp
/*
 * This is a sample program to run PaddleOCR-VL-1.5 vision-language inference in OpenCV
 * using ONNX models and OpenCV's new DNN engine (ENGINE_NEW). Given an image and a text
 * prompt, it generates a text response (e.g. recognized document text).
 *
 * The model is split into three ONNX files:
 *     - Vision encoder : image patches -> image embeddings
 *     - Embedding      : prompt token ids -> text embeddings
 *     - Decoder        : [image embeddings | text embeddings] -> logits (with KV-cache)
 *
 * Model: https://huggingface.co/PaddlePaddle/PaddleOCR-VL
 * ONNX:  https://huggingface.co/onnx-community/PaddleOCR-VL-1.5-ONNX
 *
 * Run the sample:
 * 1. Download the plain (non-quantized) fp32 ONNX export into <model_dir>, keeping the
 *    upstream layout (config.json, tokenizer.json, processor_config.json at the root;
 *    onnx/vision_encoder.onnx, onnx/embedding.onnx, onnx/decoder.onnx under onnx/).
 * 2. Run:
 *
 *      ./paddleocr_vl_inference --model_dir=<model_dir> \
 *                               --input=<path-to-image>
 */

#include <algorithm>
#include <cmath>
#include <cstring>
#include <iostream>
#include <sstream>
#include <string>
#include <vector>

#include <opencv2/core.hpp>
#include <opencv2/dnn.hpp>
#include <opencv2/imgcodecs.hpp>
#include <opencv2/imgproc.hpp>

using namespace cv;
using namespace cv::dnn;
using namespace std;

static int configValue(const FileStorage& config, const string& name, int fallback)
{
    FileNode node = config[name];
    if (!node.empty())
        return (int)node;
    FileNode textConfig = config["text_config"];
    if (!textConfig.empty() && !textConfig[name].empty())
        return (int)textConfig[name];
    return fallback;
}

static void smartResize(int height, int width, int factor, int minPixels, int maxPixels,
                        int& outHeight, int& outWidth)
{
    if (height < factor)
    {
        width = (int)round((double)(width * factor) / height);
        height = factor;
    }
    if (width < factor)
    {
        height = (int)round((double)(height * factor) / width);
        width = factor;
    }
    CV_CheckLE((double)max(height, width) / min(height, width), 200.0, "absolute aspect ratio is too large");

    int hBar = (int)round((double)height / factor) * factor;
    int wBar = (int)round((double)width / factor) * factor;
    if ((int64_t)hBar * wBar > maxPixels)
    {
        double beta = sqrt((double)(height * width) / maxPixels);
        hBar = (int)(floor(height / beta / factor)) * factor;
        wBar = (int)(floor(width / beta / factor)) * factor;
    }
    else if ((int64_t)hBar * wBar < minPixels)
    {
        double beta = sqrt((double)minPixels / (height * width));
        hBar = (int)(ceil(height * beta / factor)) * factor;
        wBar = (int)(ceil(width * beta / factor)) * factor;
    }
    outHeight = hBar;
    outWidth = wBar;
}

static Mat preprocessImage(const Mat& imageBgr, int patchSize, int mergeSize, int minPixels, int maxPixels,
                           const Vec3f& mean, const Vec3f& std_, float rescaleFactor, int& gridH, int& gridW)
{
    int factor = patchSize * mergeSize;
    int resizedHeight, resizedWidth;
    smartResize(imageBgr.rows, imageBgr.cols, factor, minPixels, maxPixels, resizedHeight, resizedWidth);

    Mat resized;
    resize(imageBgr, resized, Size(resizedWidth, resizedHeight), 0, 0, INTER_CUBIC);
    Mat rgb;
    cvtColor(resized, rgb, COLOR_BGR2RGB);
    rgb.convertTo(rgb, CV_32F, rescaleFactor);

    vector<Mat> channels(3);
    split(rgb, channels);
    for (int c = 0; c < 3; c++)
        channels[c].convertTo(channels[c], -1, 1.0 / std_[c], -mean[c] / std_[c]);

    gridH = resizedHeight / patchSize;
    gridW = resizedWidth / patchSize;
    int numPatches = gridH * gridW;

    int sizes[] = {1, numPatches, 3, patchSize, patchSize};
    Mat pixelValues(5, sizes, CV_32F);

    int idx = 0;
    for (int h = 0; h < gridH; h++)
        for (int w = 0; w < gridW; w++)
        {
            Rect roi(w * patchSize, h * patchSize, patchSize, patchSize);
            for (int c = 0; c < 3; c++)
            {
                Mat dst(patchSize, patchSize, CV_32F, pixelValues.ptr<float>(0, idx, c));
                channels[c](roi).copyTo(dst);
            }
            idx++;
        }
    return pixelValues;
}

static string buildPrompt(const string& prompt, int imageTokenRepeats)
{
    ostringstream oss;
    oss << "<|begin_of_sentence|>User: <|IMAGE_START|>";
    for (int i = 0; i < imageTokenRepeats; i++)
        oss << "<|IMAGE_PLACEHOLDER|>";
    oss << "<|IMAGE_END|>" << prompt << "\nAssistant:\n";
    return oss.str();
}

static int argmaxLastToken(const Mat& logits)
{
    int seqLen = logits.size[1];
    int vocabSize = logits.size[2];
    const float* row = logits.ptr<float>(0, seqLen - 1);
    return (int)(max_element(row, row + vocabSize) - row);
}

static vector<int> paddleocrVlInference(Net& visionNet, Net& embedNet, Net& decoderNet,
                                        const Mat& pixelValues, const Mat& imageGridThw,
                                        const string& prompt, int maxNewTokens, Tokenizer& tokenizer,
                                        int imageTokenId, int eosTokenId, int mergeSize)
{
    cout << "Inferencing PaddleOCR-VL-1.5 model..." << endl;

    int64_t gridT = imageGridThw.at<int64_t>(0, 0);
    int64_t gridH = imageGridThw.at<int64_t>(0, 1);
    int64_t gridW = imageGridThw.at<int64_t>(0, 2);
    int imageTokenRepeats = (int)((gridT * gridH * gridW) / mergeSize / mergeSize);

    vector<int> tokens = tokenizer.encode(buildPrompt(prompt, imageTokenRepeats));
    int promptLen = (int)tokens.size();
    vector<int64_t> inputIdsData(tokens.begin(), tokens.end());
    int idsShape[] = {1, promptLen};
    Mat inputIds(2, idsShape, CV_64S, inputIdsData.data());

    visionNet.setInput(pixelValues, "pixel_values");
    visionNet.setInput(imageGridThw, "image_grid_thw");
    Mat imageEmbeds = visionNet.forward();

    embedNet.setInput(inputIds, "input_ids");
    Mat inputsEmbeds = embedNet.forward();

    int hiddenDim = inputsEmbeds.size[2];
    float* embedsData = inputsEmbeds.ptr<float>();
    const float* featData = imageEmbeds.ptr<float>();
    int featIdx = 0;
    for (int i = 0; i < promptLen; i++)
        if (tokens[i] == imageTokenId)
            memcpy(embedsData + (size_t)i * hiddenDim, featData + (size_t)(featIdx++) * hiddenDim, hiddenDim * sizeof(float));

    decoderNet.enableKVCache();
    vector<int64_t> maskData(promptLen, 1);
    int attnShape[] = {1, promptLen};
    Mat attentionMask(2, attnShape, CV_64S, maskData.data());

    decoderNet.setInput(inputsEmbeds, "inputs_embeds");
    decoderNet.setInput(attentionMask, "attention_mask");
    Mat logits = decoderNet.forward();
    int newId = argmaxLastToken(logits);
    vector<int> generated = {newId};

    for (int step = 0; step < maxNewTokens - 1; step++)
    {
        if (newId == eosTokenId)
            break;

        int64_t idData[1] = {newId};
        int idShape[] = {1, 1};
        Mat newIdMat(2, idShape, CV_64S, idData);
        embedNet.setInput(newIdMat, "input_ids");
        Mat newEmbed = embedNet.forward();

        maskData.push_back(1);
        int newAttnShape[] = {1, (int)maskData.size()};
        Mat newAttentionMask(2, newAttnShape, CV_64S, maskData.data());

        decoderNet.setInput(newEmbed, "inputs_embeds");
        decoderNet.setInput(newAttentionMask, "attention_mask");
        logits = decoderNet.forward();
        newId = argmaxLastToken(logits);
        generated.push_back(newId);
    }

    if (!generated.empty() && generated.back() == eosTokenId)
        generated.pop_back();

    return generated;
}

int main(int argc, char** argv)
{
    const string keys =
        "{ help h         |      | Print help message }"
        "{ model_dir      |      | Path to the local onnx-community/PaddleOCR-VL-1.5-ONNX export "
        "(config.json, processor_config.json, "
        "onnx/vision_encoder.onnx, onnx/embedding.onnx, onnx/decoder.onnx) }"
        "{ input i        |      | Path to the input image }"
        "{ prompt         | OCR | Task prompt }"
        "{ max_new_tokens | 512  | Maximum number of new tokens to generate }"
        "{ seed           | 0    | Random seed }";

    CommandLineParser parser(argc, argv, keys);
    parser.about("Use this sample to run PaddleOCR-VL-1.5 vision-language inference in OpenCV");
    if (parser.has("help") || !parser.has("model_dir") || !parser.has("input"))
    {
        parser.printMessage();
        return 0;
    }

    string modelDir = parser.get<String>("model_dir");
    string inputPath = parser.get<String>("input");
    string prompt = parser.get<String>("prompt");
    int maxNewTokens = parser.get<int>("max_new_tokens");
    setRNGSeed(parser.get<int>("seed"));

    cout << "Preparing PaddleOCR-VL-1.5 model..." << endl;
    Tokenizer tokenizer = Tokenizer::loadVLM(modelDir + "/", "paddleocr-vl");

    FileStorage config(modelDir + "/config.json", FileStorage::READ | FileStorage::FORMAT_JSON);
    if (!config.isOpened())
    {
        cerr << "Could not open " << modelDir + "/config.json" << endl;
        return 1;
    }

    FileStorage processorFs(modelDir + "/processor_config.json", FileStorage::READ | FileStorage::FORMAT_JSON);
    if (!processorFs.isOpened())
    {
        cerr << "Could not open " << modelDir + "/processor_config.json" << endl;
        return 1;
    }
    FileNode preprocessor = processorFs["image_processor"];

    int imageTokenId = configValue(config, "image_token_id", 0);
    int eosTokenId = configValue(config, "eos_token_id", 2);

    int patchSize = preprocessor["patch_size"].empty() ? 14 : (int)preprocessor["patch_size"];
    int mergeSize = preprocessor["merge_size"].empty() ? 2 : (int)preprocessor["merge_size"];
    int minPixels = preprocessor["min_pixels"].empty() ? 28 * 28 * 130 : (int)preprocessor["min_pixels"];
    int maxPixels = preprocessor["max_pixels"].empty() ? 28 * 28 * 1280 : (int)preprocessor["max_pixels"];
    float rescaleFactor = preprocessor["rescale_factor"].empty() ? 1.0f / 255.0f : (float)preprocessor["rescale_factor"];

    Vec3f mean(0.5f, 0.5f, 0.5f), stdDev(0.5f, 0.5f, 0.5f);
    if (!preprocessor["image_mean"].empty())
    {
        vector<float> meanVec, stdVec;
        preprocessor["image_mean"] >> meanVec;
        preprocessor["image_std"] >> stdVec;
        mean = Vec3f(meanVec[0], meanVec[1], meanVec[2]);
        stdDev = Vec3f(stdVec[0], stdVec[1], stdVec[2]);
    }

    Net visionNet  = readNetFromONNX(modelDir + "/onnx/vision_encoder.onnx", ENGINE_NEW);
    Net embedNet   = readNetFromONNX(modelDir + "/onnx/embedding.onnx", ENGINE_NEW);
    Net decoderNet = readNetFromONNX(modelDir + "/onnx/decoder.onnx", ENGINE_NEW);

    Mat image = imread(inputPath);
    if (image.empty())
    {
        cerr << "Could not read image: " << inputPath << endl;
        return 1;
    }

    int gridH, gridW;
    Mat pixelValues = preprocessImage(image, patchSize, mergeSize, minPixels, maxPixels, mean, stdDev, rescaleFactor, gridH, gridW);
    int gridShape[] = {1, 3};
    Mat imageGridThw(2, gridShape, CV_64S);
    imageGridThw.at<int64_t>(0, 0) = 1;
    imageGridThw.at<int64_t>(0, 1) = gridH;
    imageGridThw.at<int64_t>(0, 2) = gridW;

    cout << "Prompt:\n" << prompt << endl;
    vector<int> generated = paddleocrVlInference(visionNet, embedNet, decoderNet, pixelValues, imageGridThw,
                                                 prompt, maxNewTokens, tokenizer, imageTokenId, eosTokenId, mergeSize);
    string response = tokenizer.decode(generated);
    cout << "Response:\n" << response << endl;
    return 0;
}
